server/liveblog/themes/themes.py

```diff
@@ -157,148 +170,286 @@ def get_options(self, theme, options=None):
         return settings
 
+    def is_local_theme(self, theme_name):
+        """
+        Check if theme is a default bundled one.
+
+        :param theme_name:
+        :return:
+        """
+        theme_folder = os.path.join(LOCAL_THEMES_DIRECTORY, theme_name)
+        return os.path.exists(theme_folder)
+
+    def is_uploaded_theme(self, theme_name):
+        """
+        Check if theme is user-uploaded.
+
+        :param theme_name:
+        :return:
+        """
+        theme_folder = os.path.join(UPLOAD_THEMES_DIRECTORY, theme_name)
+        return os.path.exists(theme_folder)
+
+    def get_theme_path(self, theme_name):
+        """
+        Get theme template path.
+
+        :param theme_name:
+        :return:
+        """
+        if self.is_local_theme(theme_name):
+            template_dir = LOCAL_THEMES_DIRECTORY
+        else:
+            template_dir = UPLOAD_THEMES_DIRECTORY
+        return os.path.join(template_dir, theme_name)
+
+    def get_theme_template_filename(self, theme_name, template_name='template.html'):
+        """
+        Get main theme template full path.
+
+        :param theme_name:
+        :param template_name:
+        :return:
+        """
+        template_path = self.get_theme_path(theme_name)
+        return os.path.join(template_path, template_name)
+
+    def get_theme_template_env(self, theme, loader=ThemeTemplateLoader):
+        """
+        Get jinja2 template environment for SEO themes.
+
+        :param theme:
+        :return:
+        """
+        embed_env = jinja2.Environment(loader=loader(theme))
+        embed_env.filters['date'] = moment_date_filter
+        return embed_env
+
+    def get_theme_compiled_templates_path(self, theme_name):
+        """
+        Get jinja2 compiled templates path for SEO themes.
+
+        :param theme_name:
+        :return:
+        """
+        compiled_templates_path = os.path.join(COMPILED_TEMPLATES_PATH, 'themes')
+        if not os.path.exists(compiled_templates_path):
+            os.makedirs(compiled_templates_path)
+
+        return os.path.join(compiled_templates_path, theme_name)
+
+    @property
+    def is_s3_storage_enabled(self):
+        # TODO: provide multiple media storage support.
+        return type(app.media).__name__ is 'AmazonMediaStorage'
+
     def get_local_themes_packages(self):
-        theme_folder = os.path.join(CURRENT_DIRECTORY, ASSETS_DIR)
-        for file in glob.glob(theme_folder + '/**/theme.json'):
-            files = []
-            for root, dirnames, filenames in os.walk(os.path.dirname(file)):
-                for filename in filenames:
-                    files.append(os.path.join(root, filename))
-            yield json.loads(open(file).read()), files
+        """
+        Get local instance theme packages.
+        It includes UPLOAD_THEMES_DIRECTORY if s3 storage is disabled.
+
+        :return:
+        """
+        theme_dirs = [LOCAL_THEMES_DIRECTORY]
+        if not self.is_s3_storage_enabled:
```

Comment:
Yes this is made to prevent to have uploaded files in the same directory where default themes are stored (as git subtree) and to prevent to keep the theme files in the local fs. This is actually breaking zip theme download